cirq/_compat.py

```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
+            setattr(self, 'load_module', loader.load_module)
+        self.old_module_name = old_module_name
+        self.new_module_name = new_module_name
+
+    def create_module(self, spec: ModuleSpec) -> ModuleType:
+        return self.loader.create_module(spec)
+
+    def module_repr(self, module: ModuleType) -> str:
+        return self.loader.module_repr(module)
+
+    def _wrap_exec_module(self, method: Any) -> Any:
+        def exec_module(module: ModuleType) -> None:
+            assert module.__name__ == self.old_module_name, (
+                f"DeprecatedModuleLoader for {self.old_module_name} was asked to "
+                f"load {module.__name__}"
+            )
+            # check for new_module whether it was loaded
+            if self.new_module_name in sys.modules:
+                # found it - no need to load the module again
+                sys.modules[self.old_module_name] = sys.modules[self.new_module_name]
+                return
+
+            # now we know we have to initialize the module
+            sys.modules[self.old_module_name] = module
+            sys.modules[self.new_module_name] = module
+
+            try:
+                return method(module)
+            except Exception as ex:
+                # if there's an error, we atomically remove both
+                del sys.modules[self.new_module_name]
+                del sys.modules[self.old_module_name]
+                raise ex
+
+        return exec_module
+
+
+def _is_internal(filename: str) -> bool:
+    """Returns whether filename is internal to python.
+
+    This is similar to how the built-in warnings module differentiates frames from internal modules.
+    """
+    return 'importlib' in filename and '_bootstrap' in filename
+
+
+_warned: Set[str] = set()
+
+
+def _deduped_module_warn_or_error(old_module_name, new_module_name, deadline):
+    if old_module_name in _warned:
+        return
+
+    _warned.add(old_module_name)
+
+    _warn_or_error(
+        f"{old_module_name} was used but is deprecated.\n "
+        f"it will be removed in cirq {deadline}.\n "
+        f"Use {new_module_name} instead.\n",
+    )
+
+
+class DeprecatedModuleFinder(importlib.abc.MetaPathFinder):
+    """A module finder to handle deprecated module references.
+
+    It sends a deprecation warning when a deprecated module is asked to be found.
+    It is meant to be used as a wrapper around existing MetaPathFinder instances.
+
+    Args:
+        finder: the finder to wrap.
+        new_module_name: the new module's fully qualified name
+        new_module_spec: the spec of the new module, that will be the root of the search path for
+            the submodules of the deprecated module
+        old_module_name: the deprecated module's fully qualified name
+        deadline: the deprecation deadline
+    """
+
+    def __init__(
+        self,
+        finder: Any,
+        new_module_name: str,
+        new_module_spec: ModuleSpec,
```

Comment:
You found the one thing I am not exactly sure about :) 
I'll have to dig into this a bit. If I just let the Finder find the spec again, I have 3 failures: 

```
FAILED cirq/_compat_test.py::test_deprecated_module[_import_multiple_deprecated-deprecation_messages5] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_b'
FAILED cirq/_compat_test.py::test_deprecated_module[_new_module_in_different_parent-deprecation_messages6] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_ops'
FAILED cirq/_compat_test.py::test_deprecated_module[_import_top_level_deprecated-deprecation_messages11] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_freezegun'
=
```

Let me think this through, most of the other comments are addressed.

Reply:
I suspect this is because the `path` (and `target`) arguments to `finder.find_spec` need to be adjusted for the new parent module (https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec).

Try
```python
        # find the corresponding spec in the new structure
        if fullname == self.old_module_name:
            path = None
            target = None
        spec = self.finder.find_spec(
            new_fullname,
            path=path,
            target=target,
        )
```

Reply:
You are right, this does boil down to updating path - and maybe target, though I haven't seen any Finder yet that uses target.

E.g. this works:

```
      if fullname == self.old_module_name:
            spec = self.finder.find_spec(
                new_fullname,
                path=["/Users/balintp/dev/proj/Cirq/cirq/testing/_compat_test_data/module_a",
                      "/Users/balintp/dev/proj/Cirq/cirq/testing/_compat_test_data",
                      "/Users/balintp/dev/proj/Cirq/cirq",
                      "/Users/balintp/dev/proj/Cirq/cirq/testing",
                      ] + sys.path,
                target=None,
            )
```

However, finding the parent module's path is a harder problem than it looks. 
Finders are not smart enough to be able to figure the parent path by themselves. 

Based on https://github.com/python/cpython/blob/master/Lib/importlib/util.py#L92-L103 this also works:


```
        # find the corresponding spec in the new structure
        if fullname == self.old_module_name:
            # this is the first time the deprecated module is being found
            # which means that the new parent needs to be found first and under
            # the new parent's path, we should be able to find the new name of
            # the deprecated module
            # this code is heavily inspired by importlib.util.find_spec
            parent_name = new_fullname.rpartition('.')[0]
            if parent_name:
                parent = __import__(parent_name, fromlist=['__path__'])
                # note that compared to importlib.util.find_spec we don't handle
                # AttributeError here because it is not expected to happen in case
                # of a DeprecatedModuleLoader - the new parent should exist and be
                # a proper package
                parent_path = parent.__path__
            else:
                parent_path = None
            spec = self.finder.find_spec(new_fullname, parent_path, None)
        else:
            # we are finding a submodule of the parent of the deprecated module,
            # which means that the parent was already found, and thus, `path` is
            # correctly pointing to the module's parent in the new hierarchy
            spec = self.finder.find_spec(
                new_fullname,
                path=path,
                target=target,
            )
```

It's a bit more complicated, but it is cleaner, more principled. I'll update it :)